[1.12] Add a World argument to WorldtypeEvent

maruohon opened the pull request:
This adds a World argument to WorldTypeEvent.BiomeSize and WorldTypeEvent.InitBiomeGens, so that those events can be targeted to specific dimensions by mods.

The argument is optional (Nullable), because the BiomeProvider is sometimes created without access to a World, such as from the AnvilSaveConverter when converting pre-Anvil worlds to Anvil format.

Offtopic:
The patch still looks a bit funky because the constructor/method bodies changed to different methods. How should these be handled? Should those params be named to their srg names to keep the patches smaller?

mezz commented:
I think you can fix up the patches by adding your new methods to the `forge.exc` file.

maruohon committed:
Add a World argument to WorldtypeEvent

This adds a World argument to WorldTypeEvent.BiomeSize and
WorldTypeEvent.InitBiomeGens, so that those events can be
targeted to specific dimensions by mods.

The argument is optional (Nullable), because the BiomeProvider is
sometimes created without access to a World, such as from the
AnvilSaveConverter when converting pre-Anvil worlds to Anvil format.

maruohon commented:
I now added the lines to `forge.exc` and the extra patch noise went away.

mezz commented on patches/minecraft/net/minecraft/world/biome/BiomeProvider.java.patch:
```diff
@@ -14,21 +14,30 @@
      }
  
-+    public GenLayer[] getModdedBiomeGenerators(WorldType worldType, long seed, GenLayer[] original)
++    public GenLayer[] getModdedBiomeGenerators(@javax.annotation.Nullable net.minecraft.world.World world, WorldType worldType, long seed, GenLayer[] original)
```
Public methods like this can't be changed directly, they need to be deprecated and replaced with new methods.

mezz commented on patches/minecraft/net/minecraft/world/gen/layer/GenLayer.java.patch:
```diff
@@ -39,9 +48,9 @@
 +    }
 +
-+    public static int getModdedBiomeSize(WorldType worldType, int original)
++    public static int getModdedBiomeSize(@javax.annotation.Nullable net.minecraft.world.World world, WorldType worldType, int original)
```
Public methods like this can't be changed directly, they need to be deprecated and replaced with new methods.

mezz commented on src/main/java/net/minecraftforge/event/terraingen/WorldTypeEvent.java:
```diff
@@ -19,31 +19,43 @@
     private final WorldType worldType;
 
-    public WorldTypeEvent(WorldType worldType)
+    public WorldTypeEvent(@Nullable World world, WorldType worldType)
```
Public methods like this can't be changed directly, they need to be deprecated and replaced with new methods.

mezz commented on src/main/java/net/minecraftforge/event/terraingen/WorldTypeEvent.java:
```diff
@@ -69,9 +82,9 @@ public WorldType getWorldType()
         private int newSize;
 
-        public BiomeSize(WorldType worldType, int original)
+        public BiomeSize(@Nullable World world, WorldType worldType, int original)
```
Public methods like this can't be changed directly, they need to be deprecated and replaced with new methods.

mezz commented on src/main/java/net/minecraftforge/event/terraingen/WorldTypeEvent.java:
```diff
@@ -114,9 +128,9 @@ public void setNewSize(int newSize)
         private GenLayer[] newBiomeGens;
 
-        public InitBiomeGens(WorldType worldType, long seed, GenLayer[] original)
+        public InitBiomeGens(@Nullable World world, WorldType worldType, long seed, GenLayer[] original)
```
Public methods like this can't be changed directly, they need to be deprecated and replaced with new methods.

mezz reviewed (changes requested):
Looks pretty good but needs to avoid breaking changes.
